dd-java-agent/instrumentation/hystrix-1.4/src/main/java/datadog/trace/instrumentation/hystrix/HystrixInstrumentation.java

```
@@ -0,0 +1,259 @@
+        }
+      } else {
+        delegate.onStart();
+      }
+    }
+
+    @Override
+    public void onNext(final T value) {
+      final Span span = spanRef.get();
+      if (span != null) {
+        try (final Scope scope = scopeManager.activate(span, false)) {
+          if (scope instanceof TraceScope) {
+            ((TraceScope) scope).setAsyncPropagation(true);
+          }
+          delegate.onNext(value);
+        } catch (final Throwable e) {
+          onError(e);
+        }
+      } else {
+        delegate.onNext(value);
+      }
+    }
+
+    @Override
+    public void onCompleted() {
+      final Span span = spanRef.getAndSet(null);
+      if (span != null) {
+        boolean errored = false;
+        try (final Scope scope = scopeManager.activate(span, false)) {
+          if (scope instanceof TraceScope) {
+            ((TraceScope) scope).setAsyncPropagation(true);
+          }
+          delegate.onCompleted();
+        } catch (final Throwable e) {
+          // Repopulate the spanRef for onError
+          spanRef.compareAndSet(null, span);
+          onError(e);
+          errored = true;
+        } finally {
+          // finish called by onError, so don't finish again.
+          if (!errored) {
+            DECORATE.beforeFinish(span);
+            span.finish();
+          }
+        }
+      } else {
+        delegate.onCompleted();
+      }
+    }
+
+    @Override
+    public void onError(final Throwable e) {
+      final Span span = spanRef.getAndSet(null);
+      if (span != null) {
+        try (final Scope scope = scopeManager.activate(span, false)) {
+          if (scope instanceof TraceScope) {
+            ((TraceScope) scope).setAsyncPropagation(true);
+          }
+          DECORATE.onError(span, e);
+          delegate.onError(e);
+        } catch (final Throwable e2) {
+          DECORATE.onError(span, e2);
+          // This recursive call might be dangerous... not sure what the best response is.
+          onError(e2);
```

Are we sure we need this? In the `try` the exception is most probably caused, at least in theory, from `delegate.onError(e)` (which is the original one). Why are we recalling the origin instead of re-throwing the exception upward?

---

Yeah, I wasn't sure about this.  You think we should just rethrow the error then?

---

I think it makes sense. We just use the hook to correctly set metadata on the span which is our end goal. I think this is good.